File: native/src/blocking/cache.rs
//! Bounded, namespace-isolated blocking signature cache.
// SPDX-License-Identifier: LGPL-3.0-or-later

use std::collections::BTreeMap;

use super::model::{BlockingControl, BlockingError, CoreBlockingMode, DirectCheckerKind};
use super::projection::BlockingSignature;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BlockingCacheNamespace {
    pub ontology_fingerprint: String,
    pub vocabulary_fingerprint: String,
    pub checker_kind: DirectCheckerKind,
    pub core_mode: CoreBlockingMode,
    pub configuration_fingerprint: String,
}

impl BlockingCacheNamespace {
    pub fn new(
        ontology_fingerprint: impl Into<String>,
        vocabulary_fingerprint: impl Into<String>,
        checker_kind: DirectCheckerKind,
        core_mode: CoreBlockingMode,
        configuration_fingerprint: impl Into<String>,
    ) -> Result<Self, BlockingError> {
        let value = Self {
            ontology_fingerprint: ontology_fingerprint.into(),
            vocabulary_fingerprint: vocabulary_fingerprint.into(),
            checker_kind,
            core_mode,
            configuration_fingerprint: configuration_fingerprint.into(),
        };
        if value.ontology_fingerprint.is_empty()
            || value.vocabulary_fingerprint.is_empty()
            || value.configuration_fingerprint.is_empty()
        {
            return Err(BlockingError::invalid(
                "blocking cache fingerprints must be nonempty",
            ));
        }
        Ok(value)
    }

    #[must_use]
    pub fn key(&self) -> (&str, &str, &'static str, &'static str, &str) {
        (
            &self.ontology_fingerprint,
            &self.vocabulary_fingerprint,
            self.checker_kind.as_str(),
            self.core_mode.as_str(),
            &self.configuration_fingerprint,
        )
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct CachePromotionContext {
    pub satisfiable: bool,
    pub completed: bool,
    pub has_nominals: bool,
    pub has_additional_ontology: bool,
    pub query_local_axioms: bool,
    pub aborted: bool,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct CachePromotion {
    pub inserted: usize,
    pub entry_count: usize,
    pub size_bytes: usize,
}

#[derive(Clone, Debug, Eq, PartialEq)]
struct CacheEntry {
    signature: BlockingSignature,
    size_bytes: usize,
    last_used: u64,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BlockingSignatureCache {
    namespace: BlockingCacheNamespace,
    max_entries: usize,
    max_bytes: usize,
    size_bytes: usize,
    clock: u64,
    entries: BTreeMap<Vec<u8>, CacheEntry>,
}

impl BlockingSignatureCache {
    pub fn new(
        namespace: BlockingCacheNamespace,
        max_entries: usize,
        max_bytes: usize,
    ) -> Result<Self, BlockingError> {
        if max_entries == 0 || max_bytes == 0 {
            return Err(BlockingError::invalid(
                "blocking cache bounds must be strictly positive",
            ));
        }
        Ok(Self {
            namespace,
            max_entries,
            max_bytes,
            size_bytes: 0,
            clock: 0,
            entries: BTreeMap::new(),
        })
    }

    #[must_use]
    pub const fn namespace(&self) -> &BlockingCacheNamespace {
        &self.namespace
    }

    #[must_use]
    pub fn entry_count(&self) -> usize {
        self.entries.len()
    }

    #[must_use]
    pub const fn size_bytes(&self) -> usize {
        self.size_bytes
    }

    #[must_use]
    pub const fn max_entries(&self) -> usize {
        self.max_entries
    }

    #[must_use]
    pub const fn max_bytes(&self) -> usize {
        self.max_bytes
    }

    pub fn contains(&mut self, signature: &BlockingSignature) -> Result<bool, BlockingError> {
        self.require_compatible(signature)?;
        let key = signature.canonical_bytes();
        let next = self.next_clock()?;
        let Some(entry) = self.entries.get_mut(&key) else {
            return Ok(false);
        };
        if entry.signature != *signature {
            return Err(BlockingError::invariant(
                "canonical blocking signature collision",
            ));
        }
        entry.last_used = next;
        Ok(true)
    }

    pub fn add(&mut self, signature: BlockingSignature) -> Result<bool, BlockingError> {
        self.require_compatible(&signature)?;
        let key = signature.canonical_bytes();
        let size_bytes = key.len().saturating_add(128);
        if size_bytes > self.max_bytes {
            return Ok(false);
        }
        let next = self.next_clock()?;
        if let Some(entry) = self.entries.get_mut(&key) {
            if entry.signature != signature {
                return Err(BlockingError::invariant(
                    "canonical blocking signature collision",
                ));
            }
            entry.last_used = next;
            return Ok(false);
        }
        self.entries.insert(
            key,
            CacheEntry {
                signature,
                size_bytes,
                last_used: next,
            },
        );
        self.size_bytes = self.size_bytes.saturating_add(size_bytes);
        self.evict();
        Ok(self.entries.values().any(|entry| entry.last_used == next))
    }

    pub fn promote_model<C: BlockingControl>(
        &mut self,
        signatures: impl IntoIterator<Item = BlockingSignature>,
        context: CachePromotionContext,
        control: &C,
    ) -> Result<CachePromotion, BlockingError> {
        if !context.satisfiable
            || !context.completed
            || context.has_nominals
            || context.has_additional_ontology
            || context.query_local_axioms
            || context.aborted
            || self.namespace.core_mode != CoreBlockingMode::None
        {
            return Ok(CachePromotion {
                inserted: 0,
                entry_count: self.entry_count(),
                size_bytes: self.size_bytes,
            });
        }
        let before = self.clone();
        let mut inserted = 0_usize;
        let outcome = (|| {
            for (index, signature) in signatures.into_iter().enumerate() {
                if index % 256 == 0 {
                    control.poll()?;
                }
                inserted = inserted.saturating_add(usize::from(self.add(signature)?));
                control.observe_memory(u64::try_from(self.size_bytes).unwrap_or(u64::MAX))?;
            }
            control.poll()?;
            Ok(CachePromotion {
                inserted,
                entry_count: self.entry_count(),
                size_bytes: self.size_bytes,
            })
        })();
        if outcome.is_err() {
            *self = before;
        }
        outcome
    }

    pub fn clear(&mut self) {
        self.entries.clear();
        self.size_bytes = 0;
        self.clock = 0;
    }

    #[must_use]
    pub fn fingerprints(&self) -> Vec<String> {
        let mut values = self.entries.values().collect::<Vec<_>>();
        values.sort_by_key(|entry| entry.last_used);
        values
            .into_iter()
            .map(|entry| entry.signature.sha256())
            .collect()
    }

    fn require_compatible(&self, signature: &BlockingSignature) -> Result<(), BlockingError> {
        if signature.kind != self.namespace.checker_kind {
            return Err(BlockingError::invalid(
                "blocking signature checker kind does not match cache namespace",
            ));
        }
        Ok(())
    }

    fn next_clock(&mut self) -> Result<u64, BlockingError> {
        self.clock = self
            .clock
            .checked_add(1)
            .ok_or_else(|| BlockingError::invariant("blocking cache LRU clock overflow"))?;
        Ok(self.clock)
    }

    fn evict(&mut self) {
        while self.entries.len() > self.max_entries || self.size_bytes > self.max_bytes {
            let victim = self
                .entries
                .iter()
                .min_by_key(|(key, entry)| (entry.last_used, *key))
                .map(|(key, _entry)| key.clone());
            let Some(victim) = victim else {
                self.size_bytes = 0;
                return;
            };
            if let Some(entry) = self.entries.remove(&victim) {
                self.size_bytes = self.size_bytes.saturating_sub(entry.size_bytes);
            }
        }
    }
}

File: src/pyhermit/backends/python/blocking/cache.py
# Copyright 2008, 2009, 2010 by the Oxford University Computing Laboratory
# Modifications Copyright 2026 pyHermiT contributors
# Adapted from HermiT commit 37ec30aced32ac81ebecc5e33fad255ddefcb4c3;
# see reports/licensing/adapted-files.toml.

"""Bounded soundness-gated blocking signature cache.

SPDX-License-Identifier: LGPL-3.0-or-later
"""

from __future__ import annotations

import threading
from collections import OrderedDict
from collections.abc import Iterable
from dataclasses import dataclass

from .signatures import BlockingSignature, DirectCheckerKind
from .strategy import CoreBlockingMode


@dataclass(frozen=True, slots=True)
class BlockingCacheNamespace:
    ontology_fingerprint: str
    vocabulary_fingerprint: str
    checker_kind: DirectCheckerKind
    core_mode: CoreBlockingMode = CoreBlockingMode.NONE
    configuration_fingerprint: str = "default"

    def __post_init__(self) -> None:
        for name in (
            "ontology_fingerprint",
            "vocabulary_fingerprint",
            "configuration_fingerprint",
        ):
            value = getattr(self, name)
            if not isinstance(value, str) or not value:
                raise ValueError(f"{name} must be a nonempty string")
        if not isinstance(self.checker_kind, DirectCheckerKind):
            raise TypeError("checker_kind must be DirectCheckerKind")
        if not isinstance(self.core_mode, CoreBlockingMode):
            raise TypeError("core_mode must be CoreBlockingMode")

    @property
    def key(self) -> tuple[str, str, str, str, str]:
        return (
            self.ontology_fingerprint,
            self.vocabulary_fingerprint,
            self.checker_kind.value,
            self.core_mode.value,
            self.configuration_fingerprint,
        )


class BlockingSignatureCache:
    """Thread-safe LRU; eviction changes performance only."""

    __slots__ = (
        "_bytes",
        "_entries",
        "_lock",
        "max_bytes",
        "max_entries",
        "namespace",
    )

    def __init__(
        self,
        namespace: BlockingCacheNamespace,
        *,
        max_entries: int = 4_096,
        max_bytes: int = 16 * 1024 * 1024,
    ) -> None:
        if not isinstance(namespace, BlockingCacheNamespace):
            raise TypeError("namespace must be BlockingCacheNamespace")
        for name, value in (("max_entries", max_entries), ("max_bytes", max_bytes)):
            if isinstance(value, bool) or not isinstance(value, int) or value <= 0:
                raise ValueError(f"{name} must be a positive integer")
        self.namespace = namespace
        self.max_entries = max_entries
        self.max_bytes = max_bytes
        self._entries: OrderedDict[bytes, BlockingSignature] = OrderedDict()
        self._bytes = 0
        self._lock = threading.RLock()

    def contains(self, signature: BlockingSignature) -> bool:
        self._require_compatible(signature)
        key = signature.canonical_bytes()
        with self._lock:
            known = self._entries.get(key)
            if known is None or known != signature:
                return False
            self._entries.move_to_end(key)
            return True

    def add(self, signature: BlockingSignature) -> bool:
        self._require_compatible(signature)
        key = signature.canonical_bytes()
        size = self._entry_size(key)
        if size > self.max_bytes:
            return False
        with self._lock:
            known = self._entries.get(key)
            if known is not None:
                if known != signature:
                    raise RuntimeError("canonical blocking signature collision")
                self._entries.move_to_end(key)
                return False
            self._entries[key] = signature
            self._bytes += size
            self._evict()
            return key in self._entries

    def promote_model(
        self,
        signatures: Iterable[BlockingSignature],
        *,
        satisfiable: bool,
        completed: bool,
        has_nominals: bool,
        has_additional_ontology: bool,
        query_local_axioms: bool,
        aborted: bool = False,
    ) -> int:
        flags = (
            satisfiable,
            completed,
            has_nominals,
            has_additional_ontology,
            query_local_axioms,
            aborted,
        )
        if not all(isinstance(flag, bool) for flag in flags):
            raise TypeError("cache promotion flags must be bool")
        if (
            not satisfiable
            or not completed
            or has_nominals
            or has_additional_ontology
            or query_local_axioms
            or aborted
            or self.namespace.core_mode is not CoreBlockingMode.NONE
        ):
            return 0
        inserted = 0
        for signature in signatures:
            inserted += int(self.add(signature))
        return inserted

    def clear(self) -> None:
        with self._lock:
            self._entries.clear()
            self._bytes = 0

    @property
    def entry_count(self) -> int:
        with self._lock:
            return len(self._entries)

    @property
    def size_bytes(self) -> int:
        with self._lock:
            return self._bytes

    def fingerprints(self) -> tuple[str, ...]:
        with self._lock:
            return tuple(signature.sha256 for signature in self._entries.values())

    def _require_compatible(self, signature: BlockingSignature) -> None:
        if not isinstance(signature, BlockingSignature):
            raise TypeError("signature must be BlockingSignature")
        if signature.kind is not self.namespace.checker_kind:
            raise ValueError("blocking signature checker kind does not match cache namespace")

    @staticmethod
    def _entry_size(key: bytes) -> int:
        return len(key) + 128

    def _evict(self) -> None:
        while len(self._entries) > self.max_entries or self._bytes > self.max_bytes:
            key, _signature = self._entries.popitem(last=False)
            self._bytes -= self._entry_size(key)


__all__ = ["BlockingCacheNamespace", "BlockingSignatureCache"]
